### ISSUE TITLE
Feature: Clients reinitialize after some retention period

### ISSUE THREAD

**Issue by gentledepp** [Mimetis/Dotmim.Sync#224]
Typically, sync frameworks produce a lot of tracking data that has to be cleaned up some time.
E.g. when a row is deleted, the metadata row in the _tracking table is still being kept.

This is to ensure, that clients that synchronize sometime in the future still get that deletion delivered to them.

However, at some point this metadata has to be removed from the system - otherwise the sql query time could degrade or the database could run full.

Possible solution: A configurable retention period per scope.

- in the server-side configuration a retention period can be specified (default: 4 weeks?)
- all _tracking tables of toombstoned records are only kept for 4 weeks and deleted thereafter
- each client sends the last time it synchronized a particular scope to the server
- if the server sees, that the client has last synced +4 weeks ago, it will return an error message, that the client needs to reinitialize
- the client then reinitializes the scope

Do you think this could be in the scope for Dotmim.Sync?
Basically one could implement this separately, but it feels like this could be a natural feature of Dotmim.Sync.

**Comment by Mimetis** [comment]
Yes, it's already implemented in `v0.4`

First of all, no need anymore to generate all tracking rows when initializing a new sync process.
Before, in `v0.3`, all existing rows were added to tracking tables during the initialization phase.
It was really time consuming on new big databases. 
Not needed now in `v0.4`

On the **Client** side, the option `CleanMetadatas` delete all rows in the tracking table after current sync (let's say sync **T**) for all rows from previous sync (let's say sync **T-1**) (cleary to be sure we can rollback an in progress sync **T** without loosing info from tracking table)
This options is enabled by default.

On the **Server** side, it is not automatic, by design.
You have to call a new method on `RemoteOrchestrator` called `DeleteMetadatasAsync()`
Basically, you set the tables you want to clean, through a `SyncSetup` instance, and then a `timeStampStart` long value.

```cs
var serverProvider = new SqlSyncProvider(DbHelper.GetDatabaseConnectionString(serverDbName));
var remoteOrchestrator = new RemoteOrchestrator(serverProvider);
var setup = new SyncSetup(new string[] {'Customer', 'CustomerAddress', 'Address');

// Delete all metadatas before timestamp 1000 on tables 
// 'Customer', 'CustomerAddress', 'Address'
await remoteOrchestrator.DeleteMetadatasAsync(null, setup, 1000);
```
I think a good idea should be to delete server metadatas every month on the server

*Note 1* : On the client side, if you disable the `CleanMetadatas` option, you can use the same method `DeleteMetadatasAsync()` on the `LocalOrchestrator`

*Note 2*: IF you want to get the last timestamp from you database, you can use the method `GetLocalTimestampAsync` from any `CoreProvider` instance

*Note 3*: No need to call these methods if you're using a `SqlSyncChangeTrackingProvider` provider

*Note 4*: Yep, I forgot to mention this feature in the documentation :)

**Comment by gentledepp** [comment]
Well... you are currently writing about cleaning up metadata.

Two notes from my side:
```
// Delete all metadatas before timestamp 1000 on tables 
// 'Customer', 'CustomerAddress', 'Address'
await remoteOrchestrator.DeleteMetadatasAsync(null, setup, 1000);
```
well.. normally I'd like to remove al metadata that is older than X days. How should I determine the timestamp in this case? By luck? ;-)

The important that you seem to have overlooked is, that IF metadata was deleted after 3 months from the server, clients that have not been syncing for (e.g.) 4 months will not receive the deletions from the server anymore - as those were deleted.

So in case metadata is removed from the server, you **must** be able to specify a maximum period that a client is allowed to be disconnected. Right?

So if the client was disconnected for too long, the **incremental** sync may not work any longer. 
So in that case, the client *must* do a **full resync** (or *reinitialize* as you call it I guess)

Or are we still talking about two different things?

**Comment by Mimetis** [comment]
## Correlation between timestamp and real date

You can get the current timestamp from any db using `GetLocalTimestampAsync` from any `CoreProvider` instance.

Or you can use `SELECT @@DBTS` from Sql Server

If you want to correlate this timestamp with a datetime, I assume you can make a task running every day at midnight, and save in a table the `getdate()` and `@@DBTS` value.

## Client Out of date

in `v04` we don't have a `sync_scope` table anymore.
So far, this information is not stored on the server side. (I mean the timestamp used as a limit for CleanMetadatas)
Then I'm not able to compare a last successufl timestamp from a client and make a verification if this time is still legit.

Today, you have to make this verification manually. Obviously this is NOT a good solution, and I still think about it.
I guess, I will need to re-create a `sync metadatas table` on the server side...

**Comment by gentledepp** [comment]
> I guess, I will need to re-create a sync metadatas table on the server side...
I do not understand why this is necessary honestly.
The client does track the last sync datetime, right?
Why can't it send it via another httpheader? 🙄

**Comment by Mimetis** [comment]
Let's say your client successfully sync for the last time (server time) at  `cli1_timestamp` =  **200** 
Your server has a `current_timestamp`  =  **1000**
Now, you decide to clean the server, and launch the method `DeleteMetadatasAsync` with a `limit_timestamp` arg = **500**

Then, your client launch a sync.
Basically it will send its `cli1_timestamp` to the server.
The server should check that `cli1_timestamp` (200) is >=  `limit_timestamp` (500)
In that case, the server will return an `OutOfDateException`.

The problem is that the server did not store the `limit_timestamp` value anywhere.
Does it make sense ?

If you have an idea in mind, please share !

**Comment by Mimetis** [comment]
This issue has been fixed with the release of the version `v0.5`.
If you still experiencing the same bug, please comment here, I will re-open the issue.

Regarding the `DeleteMetadataAsync` from the server side, here is a small piece of code you can run every week or two:

``` cs
 private static async Task CleanServerMetadatasAsync()
    {
        // Create the setup used for your sync process
        var tables = new string[] {"ProductCategory",
                    "ProductDescription", "ProductModel",
                    "Product", "ProductModelProductDescription",
                    "Address", "Customer", "CustomerAddress",
                    "SalesOrderHeader", "SalesOrderDetail" };

        var setup = new SyncSetup(tables)
        {
            // optional :
            StoredProceduresPrefix = "server",
            StoredProceduresSuffix = "",
            TrackingTablesPrefix = "server",
            TrackingTablesSuffix = ""
        };

        var remoteSqlSyncProvider = new SqlSyncProvider("Data Source=(localdb)\\mssqllocaldb; Initial Catalog=AdventureWorks; Integrated Security=true;");
        var remoteOrchestrator = new RemoteOrchestrator(remoteSqlSyncProvider, new SyncOptions(), setup);

        // delelete metadatas
        var dmd = await remoteOrchestrator.DeleteMetadatasAsync();

        Console.WriteLine(dmd);
    }
```
The cool thing here is that the `RemoteOrchestrator` will get, from `server_scope_hsitory` the **Min** timestamp of all clients that have already synced *at least* one time.
That way you are pretty sure that the server tracks the less amount of metadatas in the tracking tables.

Anyway, if at some point, one client is out of date, he can sync anyway, with a full reinit.
Check this sample: [OutDated](https://github.com/Mimetis/Dotmim.Sync/tree/master/Samples/OutDated) : This sample shows how to deal with an outdated client database

**Comment by gentledepp** [comment]
Hi Mimetis - thanks for the heads-up.

I see, however, one major shortcoming of this approach:

> The cool thing here is that the RemoteOrchestrator will get, from server_scope_hsitory the Min timestamp of all clients that have already synced at least one time.

Well... in the real world, this is not going to work. 
Think about this:
* A user uses your app some time with device **d1**
* then, a year later, (s)he get a new device (iPhone 77 or whatever)
* because **d1** is not used anymore, it will from now on be the "client with the lowest timestamp"
So no matter how often we call `var dmd = await remoteOrchestrator.DeleteMetadatasAsync();`, nothing is going to happen.

So I propose the following addition to your approach:

1. for each scope, allow to define a 'max_disconnection_period' after which each client will have to **reinitialize** - in our example: 14 days
2. in the table `server_scope_hsitory`, also track the current date of each entry. Then we could see that e.g. **d1** has not synchronized since more than 14 days
3. When `await remoteOrchestrator.DeleteMetadatasAsync();` is called, also delete all rows from 'server_scope_history' which were last updated longer than 14 days ago

That way you can clean up data even if some devices *never* sync again.

4. Now *if* **d1** should reconnect e.g. 21 days later (maybe the user was on a looooong holiday), it will sent its timetsamp to the server. Let's call it **d1_timestamp** and assume it is `>0` (because `0` is just a first full sync anyways)
5. The server has no more row in the `server_scope_history` table for **d1**. *However*, if there is no row for **d1** in the table, we know that it has been disconnected for too long! 💡 
This way, the server knows that **d1** has to reinitialize. Otherwise it may miss some important data changes.
6. so in that case, the server ignores **d1_timestamp** and uses the timestamp 0.
7. Also, in the response to the client, it will add the `reinitialize` flag. (I guess you could put this in the response message)
8. The client will receive the data and notice the `reinitialize` flag. It will therefore cleanup all local data and apply the changes coming from the server

This would be a very elegant solution. What do you think?

Another benefit of having the last sync date in the `server_scope_history` table would be that we know when the device last synchronized.
I can imagine another nice scenario, where we add the parameters of a sync (in our case UserId) to the `server_scope_history` table. That would allow us to see when **each user** synchronized and * how many devices* they have. ;-)

**Comment by Mimetis** [comment]
Basically, the only problem with your proposition (but I guess we can find a workaround) is that the `timestamp` used to mark any row, is **not** a datetime related value but **just** a long value increasing each time a row is inserted / deleted / updated

> From Sql docs
> The SQL Server timestamp data type has nothing to do with times or dates. SQL Server timestamps are binary numbers that indicate the relative sequence in which data modifications took place in a database. The timestamp data type was originally implemented to support the SQL Server recovery algorithms.

## Theory

On the server side, in `scope_info_server`, we have the `sync_scope_last_clean_timestamp` column, who contains the *last timestamp* used when calling `DeleteMetadatasAsync()`

(*I removed irrelevant columns for this explanation*)
``` sql
SELECT * [sync_scope_name] ,[sync_scope_schema], [sync_scope_last_clean_timestamp]
FROM [AdventureWorks].[dbo].[scope_info_server]
```
**Server database:**
sync_scope_name | sync_scope_schema | sync_scope_last_clean_timestamp
------------- | ----------| ------------
DefaultScope | "{......}" | **1800**

Here we have a last timestamp value of **1800**

So far, any client who tries to make a sync will use its own `scope_last_server_sync_timestamp` from the client `scope_info` table.

``` sql
SELECT [sync_scope_id] ,[sync_scope_name] ,[scope_last_server_sync_timestamp]
FROM [Client].[dbo].[scope_info]
```
**Client database:**
sync_scope_id | sync_scope_name | scope_last_server_sync_timestamp
------------- | ----------| ------------
9E9722CD-... | DefaultScope  | **2000**

So far, **2000** from client db is superior to **1800** from server db. Everything is fine.

Whatever a line exists, or not, in server `scope_info_history`, it's not used in the flow (Eventually, this table is just for information purpose)

## From scratch
Ok, now let's see what happens "from scratch"

Imagine we *never* made any call to `DeleteMetadatasAsync()`, since the begining.   
We may have something like this, in the `scope_info_server` table:

**Server database:**
sync_scope_name | sync_scope_schema | sync_scope_last_clean_timestamp
------------- | ----------| ------------ 
DefaultScope | "{......}" | **0**

And we have this kind of lines in `sync_scope_history`, assuming some clients have made some synchronization:

``` sql
SELECT [sync_scope_id] ,[sync_scope_name] ,[scope_last_sync_timestamp], [scope_last_sync]
FROM [AdventureWorks].[dbo].[scope_info_history]
```
**Server database:**
sync_scope_id | sync_scope_name | scope_last_sync_timestamp | scope_last_sync
------------- | ----------| ------------ | -----------------
9E9722CD-... | DefaultScope | **10** | 2018-07-01
AB4122AE-... | DefaultScope | **2100** | 2020-04-10
DB6EEC7E-... | DefaultScope | **2000** | 2020-03-20
E9CBB51D-... | DefaultScope | **2020** | 2020-03-21
CC8A9184-... | DefaultScope | **2030** | 2020-03-22
D789288E-... | DefaultScope | **2040** | 2020-03-23
95425970-... | DefaultScope | **2050** | 2020-03-24
5B6ACCC0-... | DefaultScope | **2060** | 2020-03-25

So far, as you say, if we call the `DeleteMetadataAsync()` on the server side, we will delete only metadatas rows where `scope_last_sync_timestamp < 10` ...
As you said, not ideal.

Then I guess we can make something like that:

``` cs
// get all history lines
var histories = await remoteOrchestrator.GetServerHistoryScopes();

// select only clients that have synced at least 30 days earlier
var historiesTwoWeeksAgo = histories.Where(h => h.LastSync.HasValue && h.LastSync.Value >= DateTime.Now.AddDays(-30));

// Get the min timestamp
var minTimestamp = historiesTwoWeeksAgo.Min(h => h.LastSyncTimestamp);

// Call the delete metadatas with this timestamp
await remoteOrchestrator.DeleteMetadatasAsync(minTimestamp);
```

With this piece of code, I guess the min timestamp retrieved will be **2100**.
Then the first client became *out of date* and the next time he will sync, he will have to make a full reinitialization.

Your thought ?

**Comment by gentledepp** [comment]
> Basically, the only problem with your proposition (but I guess we can find a workaround) is that the timestamp used to mark any row, is not a datetime related value but just a long value increasing each time a row is inserted / deleted / updated

Well...I know. My proposition was to introduce a `scope_last_sync` column - just as you show here ;-)

However, your code is wrong:
All clients **must** sync within 30 days.
So all changes **older than 30 days** can be deleted.

=> 
```
// get all history lines
var histories = await remoteOrchestrator.GetServerHistoryScopes();

// select only clients that have synced at least 30 days earlier
var historiesTwoWeeksAgo = histories.Where(h => h.LastSync.HasValue && h.LastSync.Value >= DateTime.Now.AddDays(-30));

// Get the MAX timestamp - because ALL those timestamps are out of the valid time period already!!
var minTimestamp = historiesTwoWeeksAgo.Max(h => h.LastSyncTimestamp);

// Call the delete metadatas with this timestamp
await remoteOrchestrator.DeleteMetadatasAsync(minTimestamp);

``` 

However, I would *urge* to do this with a SQL query, and not in-memory, because there can be quite some clients (several thousands) and doing this in-memory is unnecessarily inefficient:

getting the `minTimestamp`
```
declare @thirtyDaysAgo datetime2(7);

SELECT max([scope_last_sync_timestamp]) 
FROM [AdventureWorks].[dbo].[scope_info_history]
WHERE [scope_last_sync] < @thirtyDaysAgo
```

once the cleanup succeeded (and only if so), clean those outdated rows from the `[scope_info_history]` table.
```
declare @thirtyDaysAgo datetime2(7);

Delete FROM [AdventureWorks].[dbo].[scope_info_history]
WHERE [scope_last_sync] < @thirtyDaysAgo
```

Because if you delete the history records and `await remoteOrchestrator.DeleteMetadatasAsync(minTimestamp);` **fails**, the obsolete metadata may remain in the database for quite a while!


Are we now speaking the same language? :-)

**Comment by Mimetis** [comment]
Yes, we are basically saying the same thing :)

I'm not sure we have the same conclusions on the `historiesTwoWeeksAgo`:

``` cs
var historiesTwoWeeksAgo = histories.Where(h => h.LastSync.HasValue && h.LastSync.Value >= DateTime.Now.AddDays(-30));
```
This line gets **Valid** clients (who LastSync occured during **Now** and **Now-30 days**   
So far, the timestamp we should consider is the **Min** timestamp of these valid clients, right ?

I agree that making it from within SQL with a query may be a better performance solution.
I will make a PR for a next version.

Thanks for the review. Let's figure out together what is the correct algorithm and I will create the doc page, right after.

**Comment by gentledepp** [comment]
uups.... apoligies.. I read that incorrectly.
Sry for that.

So then **yes** I agree :-)

**Comment by Mimetis** [comment]
Doc created for Metdata rows  management: https://mimetis.github.io/Dotmim.Sync/Metadatas

**Comment by gentledepp** [comment]
Nice work! 
I have to say, though, that I liked the old theme of your docs website much better.
The red color of the headlines looks as if there was an error.
But mostly: Have a look at it from your mobile - the layout is terrible :-)

**Comment by Mimetis** [comment]
Yes, I know, it's pretty ugly :D
I'm working on it. I'm not really good at css & styling, but It should be a little bit better right now :)
